_source/_posts/2019-09-24-how-to-create-a-symfony-app-with-authentication.md

```diff
@@ -0,0 +1,771 @@
+
+Once you've created your `User` entity, you can configure the Security bundle so it knows where the users are stored (in the database, in memory, using HTTP basic authorization, etc.), what algorithm to use for encoding passwords, and various other options.
+
+The most important section of the security configuration file is `firewalls`. You can have multiple firewalls, but only one is active during each request, and Symfony determines it by finding the first firewall that matches the URL pattern. The firewall defines if a user should be authenticated to access the requested resource, what user roles are required, etc.
+
+I'll review the Symfony security system in more detail (and in a more practical way) during the tutorial.
+
+## Okta Registration and Application Setup
+
+Before proceeding with the Symfony application, set up your Okta account and application. Okta makes [identity management](https://developer.okta.com/product/user-management/) easier, more secure, and more scalable than you're used to. Okta is an API service that allows you to create, edit, and securely store user accounts and user account data, and connect them with one or more applications.
+
+[Register for a forever-free developer account](https://developer.okta.com/signup/), then log in to your Okta account, go to the Applications menu item and click the **Add Application** button:
+
+{% img blog/symfony-authentication/add-application.png alt:"Add Application button" width:"260" %}{: .center-image }
+  
+Select **Web** and click **Next**.
+
+{% img blog/symfony-authentication/choose-web-app.png alt:"Select the application type" width:"800" %}{: .center-image }
+
+Enter a title, and set `http://localhost:8000/` as the Base URI, and `http://localhost:8000/authorization-code/callback` as the Login Redirect URI, then click **Done**. You can leave the rest of the settings as they are.
+
+Copy the Client ID and Client Secret from the application settings, you will need them soon.
+
+## Set up the Symfony Project
+
+Create a MySQL database and user:
+
+```bash
+mysql -uroot -p
+CREATE DATABASE symfony_auth CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
+GRANT ALL on symfony_auth.* to 'symfony_auth'@'127.0.0.1' identified by 'symfony_auth';
+quit
+```
+
+Initialize a new Symfony project using the Flex skeleton and start the built-in PHP server:
+
+```bash
+composer create-project symfony/skeleton symfony-auth-demo
+cd symfony-auth-demo
+php -S 127.0.0.1:8000 -t public
+```
+
+Load `http://127.0.0.1:8000` and you should see the default Symfony page.
+
+Make sure the version of Symfony is locked to 4.3 in `composer.json` by confirming the versions match the below:
+
+```
+...
+        "symfony/console": "4.3.*",
+        "symfony/dotenv": "4.3.*",
+        "symfony/flex": "^1.3.1",
+        "symfony/framework-bundle": "4.3.*",
+        "symfony/yaml": "4.3.*"
+...
```

Comment:
I typically left-align code blocks, but I'll leave this up to you.

Reply:
I'll leave this since it matches the indentation of the file in case someone copypastes it.